lock-impl/src/main/java/com/palantir/lock/logger/LockServiceStateLogger.java

```diff
@@ -0,0 +1,216 @@
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.DumperOptions;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.nodes.Tag;
+import org.yaml.snakeyaml.representer.Representer;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimaps;
+import com.google.common.collect.SetMultimap;
+import com.palantir.lock.HeldLocksToken;
+import com.palantir.lock.LockClient;
+import com.palantir.lock.LockRequest;
+import com.palantir.lock.impl.LockServiceImpl;
+
+public class LockServiceStateLogger {
+
+    private static Logger log = LoggerFactory.getLogger(LockServiceStateLogger.class);
+
+    private static final String YAML_FILE_ALREADY_EXISTS_EXCEPTION = "Yaml file either already exists or can't be created at: ";
+    private static final String WARNING_LOCK_DESCRIPTORS = "# WARNING: Lock descriptors may contain sensitive information\n";
+    private static final String FILE_NOT_CREATED_LOG_ERROR = "Destination file [{}] either already exists"
+            + "or can't be created. This is a very unlikely scenario."
+            + "Retrigger logging or check if process has permitions on the folder";
+
+
+    private static final String OUTSTANDING_LOCK_REQUESTS_TITLE = "OutstandingLockRequests";
+    private static final String HELD_LOCKS_TITLE = "HeldLocks";
+
+    private final LockDescriptorMapper lockDescriptorMapper = new LockDescriptorMapper();
+    private final long startTimestamp = System.currentTimeMillis();
+
+    private final ConcurrentMap<HeldLocksToken, LockServiceImpl.HeldLocks<HeldLocksToken>> heldLocks;
+    private final Map<LockClient, Set<LockRequest>> outstandingLockRequests;
+
+
+    public LockServiceStateLogger(ConcurrentMap<HeldLocksToken, LockServiceImpl.HeldLocks<HeldLocksToken>> heldLocksTokenMap,
+            SetMultimap<LockClient, LockRequest> outstandingLockRequestMultimap) {
+        this.heldLocks = heldLocksTokenMap;
+        this.outstandingLockRequests = Multimaps.asMap(outstandingLockRequestMultimap);
+    }
+
+    public void logLocks(String outputDir) throws IOException {
+        Map<String, Object> generatedOutstandingRequests = generateOutstandingLocksYaml(outstandingLockRequests);
+        Map<String, Object> generatedHeldLocks = generateHeldLocks(heldLocks);
+
+        Path outputDirPath = Paths.get(outputDir);
+        Files.createDirectories(outputDirPath);
+
+        dumpYamlsInNewFiles(outputDir, generatedOutstandingRequests, generatedHeldLocks);
+    }
+
+    private Map<String, Object> generateOutstandingLocksYaml(Map<LockClient, Set<LockRequest>> outstandingLockRequestsMap) {
+        Map<String, SimpleLockRequestsWithSameDescriptor> outstandingRequestMap = Maps.newConcurrentMap();
```

Comment:
Also can't this become Map<String, List<SimpleLockRequest>>?  And then you just sort the entries?